src/headers.js

```diff
@@ -35,68 +44,103 @@ function fill (headers, object) {
 
 class Headers {
   constructor (init) {
-    this[kHeaders] = new Map()
+    this[kHeaders] = []
 
     if (init && typeof init === 'object') {
       fill(this, init)
     }
   }
 
   append (name, value) {
-    const [normalizedHeaderName, normalizedHeaderValue] = normalizeAndValidateHeaderValue(name, value)
-
-    const existing = this[kHeaders].get(normalizedHeaderName)
-    const newHeaderValue = existing ? existing + ', ' + normalizedHeaderValue : normalizedHeaderValue
-    this[kHeaders].set(normalizedHeaderName, newHeaderValue)
+    const [normalizedName, normalizedValue] = normalizeAndValidateHeaderValue(name, value)
+
+    let index = this[kHeaders].length
+    for (let i = 0; i < this[kHeaders].length; i += 2) {
+      if (normalizedName === this[kHeaders][i]) {
+        this[kHeaders][i + 1] += `, ${normalizedValue}`
+        return
+      } else if (this[kHeaders][i] > normalizedName) {
+        index = i
+        break
+      }
+    }
+    this[kHeaders].splice(index, 0, normalizedName, normalizedValue)
```

Comment:
This kind of removes the whole point of using an array instead of a map?

Reply:
per fetch spec, the headers have to be sorted and combined when iterated over. Previous commits in this branch implemented combine-on-insert and sort-on-iterate. But sorting the internal list any time the class instance is iterated over is very slow; thus I've implemented a sort-on-insert as well.

I'll work on a benchmark script now to compare the map implementation vs the array one. My theory is that the array implementation is overall less resource intensive from both a memory and efficiency stand point. Will reply back with perf details shortly

Reply:
Another way could be to have two modes. Start with sort on iterate and then switch to sorted insert.

Reply:
```
write performance
  Map: 0.3742ms
  Array: 0.1885ms
read performance
  Map: 0.1238ms
  Array: 0.0686ms
```

https://gist.github.com/Ethan-Arrowood/1c79aee396eb6c2f1e78380b69ae0f0a

Reply:
In c++ a sorted vector is faster than a sorted map. However, from my experience that does not translate to V8.

Reply:
What I've discovered overall is that V8 has done an incredible job optimizing arrays. Even things like `...` and `.splice` are much faster than expected. The array based implementation uses as little assignments as necessary. The sort-on-insert doesn't add much to the write performance because the class was already combining-on-insert.

I'm starting to think the only faster solution here would be to _just use an array_, and then write a collection of utility header functions that can be called only when the given operation is required (i.e. normalizing, validation, combining, sorting)

Reply:
or I do all of this using WASM 🦀

Reply:
I'm inclined to give WASM a shot too, using Rust for the binary search -- we can bring it down even further. But maybe this should be postponed as a highly specialised future enhancement so you can just move on and focus on delivering a fully-compliant release first?

Reply:
yes WASM won't happen just yet; i have some other work on this library i wanna do first